src/digraphs.c

```diff
@@ -1384,6 +1384,80 @@ static bool EqJumbledPlists(Obj l, Obj r, Int nr, Int* buf) {
   return true;
 }
 
+//-----------------------------------------------------------------------------
+// MurmurHash3 was written by Austin Appleby, and is placed in the public
+// domain. The author hereby disclaims copyright to this source code.
+
+/* Minor modifications to get it to compile in C rather than C++ and
+integrate with GAP  SL*/
+
+/* Digraphs takes parts of this source from GAP */
+
+#define BIG_CONSTANT(x) (x##LLU)
+
+#ifndef SYS_IS_64_BIT
+
+static inline uint32_t fmix4(uint32_t h) {
+  h ^= h >> 16;
+  h *= 0x85ebca6b;
+  h ^= h >> 13;
+  h *= 0xc2b2ae35;
+  h ^= h >> 16;
+
+  return h;
+}
+
+#else
+
+static inline uint64_t fmix8(uint64_t k) {
+  k ^= k >> 33;
+  k *= BIG_CONSTANT(0xff51afd7ed558ccd);
+  k ^= k >> 33;
+  k *= BIG_CONSTANT(0xc4ceb9fe1a85ec53);
+  k ^= k >> 33;
+
+  return k;
+}
+
+#endif
+
+static Obj FuncDIGRAPH_HASH(Obj self, Obj digraph) {
+  UInt n, i;
+  Obj  out, a, res;
+  Int  nr, j;
+
+  n   = DigraphNrVertices(digraph);
+  out = FuncOutNeighbours(self, digraph);
+
+#ifndef SYS_IS_64_BIT
+  uint32_t* buf;
+  buf = safe_calloc(n, sizeof(uint32_t));
+#else
+  uint64_t* buf;
+  buf = safe_calloc(n, sizeof(uint64_t));
+#endif
+
+  for (i = 1; i <= n; i++) {
+    a = ELM_PLIST(out, i);
+    PLAIN_LIST(a);
+    nr = LEN_PLIST(a);
+    for (j = 1; j <= nr; j++) {
+#ifndef SYS_IS_64_BIT
+      buf[i - 1] += fmix4((uint32_t) INT_INTOBJ(ELM_PLIST(a, j)));
+#else
+      buf[i - 1] += fmix8((uint64_t) INT_INTOBJ(ELM_PLIST(a, j)));
+#endif
```

Comment:
With my suggestions from above, you could replace `buf` by a single var `UInt h = 0;` and then do
```suggestion
      h = hash_combine(h, (UInt) INT_INTOBJ(ELM_PLIST(a, j)));
```

Reply:
Thanks for the suggestion, getting rid of `buf` is very nice. I've not implemented this exact logic since we actually want the hash to be commutative on each list of out neighbors so that `Digraph([[1, 2], []])` and `Digraph([[2, 1], []])` get the same hash. I elected to do
```c
    for (j = 1; j <= nr; j++)
      h += fmix(INT_INTOBJ(ELM_PLIST(a, j)));
    h = fmix(h + 0x9e3779b9);
```
This way the order of elements within each out neighbour list does not matter, but when accumulating we introduce some non-commutativity so `Digraph([[1, 2], []])` and `Digraph([[], [1, 2]])`  should get different hashes.

Reply:
Oh indeed, if the representation is not unique then commutativity may be useful and even necessary. My apologies, I should have considered that. Thank you for doing it properly despite my misleading comment